### PR TITLE
PiLink refactor

### PR DESCRIPTION
**Pull request by speshak**
A massive refactor of the PiLink class. 

- Splits Stream handling from command processing (Introduces CommandProcessor and SettingLoader classes to take up these tasks)
- Uses C++ template to materialize differences between Stream implementations (`Serial` v. `WiFiClient`)
- Adds a ArduinoJson based interface for communicating via PiLink. (A compatibility facade is provided to bridge the gap)
- Replace String based Stream buffering with [ArduinoStreamUtils](https://github.com/bblanchon/ArduinoStreamUtils)

And some semi-unrelated changes came along for the ride as well:
- Improve/correct doxygen comments
- Refactor out some magic numbers/macros to `constexpr` values 

My goal was to improve how understandable the PiLink code is by reducing and encapsulating is various responsibilities.

### CODE REVIEW

**Comment by speshak** [comment]
I'm not 100% ready for this to be merged, but I wanted to get it out for comment.  It works great in my limited testing on the serial build.  I need to get testing done on WiFi as well.

**Comment by lbussy** [comment]
Interesting stuff - I wish there were room for some of this in the Uno code!  👍

**Comment by speshak** [comment]
And here I am wishing that the esp8266 toolchain would support C++14 😆

**Review by thorrak — COMMENTED** [review]
> And here I am wishing that the esp8266 toolchain would support C++14 😆

Don't even get me started. The lack of full C++ 11 support set TiltBridge back at least a month back when I was getting it going. Always frustrating when code that should work just... doesn't.

**Comment by thorrak** [comment]
Aaaaand my GitHub-fu is terrible. I thought those comments were being submitted as I wrote them.Sorry about that.

Still reading through everything!

File: src/CommandProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+ * Continuously reads data from PiStream and processes the command strings.
+ */
+void CommandProcessor::receiveCommand() {
+  while (piLink.available() > 0) {
+    char inByte = piLink.read();
+    switch(inByte){
+    case ' ':
+    case '\n':
+    case '\r':
+#ifdef ESP8266_WiFi  // Control characters sent when establishing a telnet session
```
[the remark this answers is not in the document]
Comment (speshak) [reply src/CommandProcessor.cpp]: Funny you should say that, I dropped those exact lines when I thought up [this](https://github.com/thorrak/brewpi-esp8266/pull/55/files#diff-4ed3154b492aa155addbee523753c9c2R53) last night.

File: .gitignore [hunk 2]
```diff
@@ -31,3 +31,5 @@ bin/esp32_wifi_tft.bin
 
 apidocs/
 *.sw?
+
+src/Version.h
```
[the remark this answers is not in the document]
Comment (speshak) [reply .gitignore:35]: If you like this one, you might also enjoy [another little idea I had](https://github.com/thorrak/brewpi-esp8266/commit/9cba22b91b9a7eb79bcaba226197c247013294fa).  That's just a proof of concept for the idea, but I think it demonstrates it fairly well.  Each build can be given a list of features to enable, and it happens all at compile time.  The only downside is that it requires C++14, and so far I've only got the toolchain updated for ESP8266.

**Comment by speshak** [comment]
I've finished testing this under the current dev branch of fermentrack.  I haven't found any regressions, so I'm going to consider this PR complete. (Or at least I'm going to stop pushing more changes to it unless an issue is identified)

**Comment by speshak** [comment]
I have ended up finding a bug I introduced to the logger as part of this.  I'll get a fix added soon.

**Pull request merged**